### DISCUSSION TITLE
Migrating from Multilingual App Toolkit - what features / workflows are you missing?

### DISCUSSION THREAD

**Discussion by tom-englert** [dotnet/ResXResourceManager#444] [General]
What features or workflows does the MAT provide that you can't adapt using RXRM?

**Comment by GeertvanHorrik** [comment]
Disclaimer: I've not yet done a PoC with ResXResourceManager (RXRM), but will explain what I like about MAT:

1. Separate *real* editor that is "easy to understand" for translators. E.g. we can install this tool (with ribbon) and people understand it. They can filter on "changed" items and thus can easily update changed / new translations.
2. Resource files are automatically generated when the resources change (and sets the status to changed, etc)
3. Machine translators (e.g. use Azure translation services)

This makes a good separation between the translators (end-users, translation companies, etc) vs programmers (that deal with resx).

What holds me back from migrating is the huge dependency we have on MAT (our translators are used to it, "t works" (but not in VS 2022) and we have many repositories to manage). If we will migrate, we will need to write some sort of migrator for ourselves (which we will of course open source as well).

We build many open source packages, so maybe we can contribute on the editor and build a similar editor as MAT which is open source and released regularly.

**Comment by tom-englert** [comment]
1. Whatever you call a "real" editor 😏. RXRM e.g. has a compact tool bar instead or a clumsy, mostly empty ribbon, editor is integrated in the grid.
2. What is the SSOT in MAT? Does it store entries also in a DB? In RXRM it's only the .resx files, it has no intermediate storage. If MAT also works with only .resx files, both could co-exist.
3. RXRM actually supports Azure, DeepL, Google, MyMemory, MS Terminology

**Reply by GeertvanHorrik** [reply]
MAT uses a separate format, see https://github.com/WildGums/Orchestra/tree/develop/src/Orchestra.Core/MultilingualResources for examples. I believe this is a standardized format for translations (XML Localisation Interchange File Format).

I think this allows MAT to manage state between the resx and the state (e.g. what has been changed and when). Maybe we can use this intermediate file for RXRM? Then people *could* use the MAT editor if they want, but also the built-in editor. But most important for translators is to be aware what must be translated, especially when sending the files for updated translators.

**Comment by tom-englert** [comment]
I see, with a real live example it becomes much clearer.
So RXRM could generate/update XLIF files when editing resources. 
But how to read updated XLIF files from translators back? Would this be a manual step by clicking some import button and selecting the files?

**Reply by GeertvanHorrik** [reply]
We send the XLIF files to our translators, and copy them back. I think a build target checks the delta by comparing the values and then updates either the RESX or XLIFF.

**Comment by tom-englert** [comment]
@GeertvanHorrik basic functionality for RESX/XLIFF synchronization is available:
- must be enabled in RXRM's configuration tab
- currently only runs when RXRM's tool window is open in VS or the standalone version is running.

There are some open questions on how to handle e.g. comments/notes, but basically it works.
Can you give it a try?

VSIX is available form VSIXGallery: https://www.vsixgallery.com/extension/43b35fe0-1f30-48de-887a-68256474202a
To try the standalone version, you must build it from the sources.

**Reply by GeertvanHorrik** [reply]
Finally got the time to look into this. MAT uses msbuild targets so I can run it automatically by including some settings.

Would something like this be possible for ResXResourceManager as well? Here is an example:

https://github.com/GeertvanHorrik/RepositoryTemplate/blob/master/src/Directory.Build.shared.mat.props

I can simply import this props file and MAT is enabled. I want to prevent developers having to explicitly open the window manually once which will, eventually, be forgotten (myself included).

**Reply by GeertvanHorrik** [reply]
@tom-englert found something: the comments seem to be removed after using this tool. For example, these items are being removed from resx after syncing:

```
  <data name="MyResource" xml:space="preserve">
    <value>You are about to start {0}.</value>
    <comment>{0} = object name</comment>
  </data>
```

gets turned into

```
  <data name="MyResource" xml:space="preserve">
    <value>You are about to start {0}.</value>
  </data>
```

**Comment by tom-englert** [comment]
comment should be fixed now.

**Comment by tom-englert** [comment]
@GeertvanHorrik latest version from [VSIXGallery](https://www.vsixgallery.com/extension/43b35fe0-1f30-48de-887a-68256474202a) now also runs a sync before build, even when the view has never been opened.
